pyomo/repn/plugins/nl_writer.py

```diff
@@ -0,0 +1,2128 @@
+    WriterFactory.register('nl', doc)(WriterFactory.get_class(f'nl_v{n}'))
+
+def categorize_valid_components(
+        model, active=True, sort=None, valid=set(), targets=set()):
+    assert active in (True, None)
+    unrecognized = {}
+    component_map = {k: [] for k in targets}
+    for block in model.block_data_objects(active=active,
+                                          descend_into=True,
+                                          sort=sort):
+        local_ctypes = block.collect_ctypes(active=None, descend_into=False)
+        for ctype in local_ctypes:
+            if ctype in kernel.base._kernel_ctype_backmap:
+                ctype = kernel.base._kernel_ctype_backmap[ctype]
+            if ctype in targets:
+                component_map[ctype].append(block)
+                continue
+            if ctype in valid:
+                continue
+            # TODO: we should rethink the definition of "active" for
+            # Components that are not subclasses of ActiveComponent
+            if not issubclass(ctype, ActiveComponent) and \
+               not issubclass(ctype, kernel.base.ICategorizedObject):
+                continue
+            if ctype not in unrecognized:
+                unrecognized[ctype] = []
+            unrecognized[ctype].extend(
+                block.component_data_objects(
+                    ctype=ctype,
+                    active=active,
+                    descend_into=False,
+                    sort=SortComponents.unsorted))
+    return component_map, {k:v for k,v in unrecognized.items() if v}
+
+@WriterFactory.register(
+    'nl_v2', 'Generate the corresponding AMPL NL file (version 2).')
+class NLWriter(object):
+    CONFIG = ConfigBlock('nlwriter')
+    CONFIG.declare('show_section_timing', ConfigValue(
+        default=False,
+        domain=bool,
+        description='Print timing after writing each section of the NL file',
+    ))
+    CONFIG.declare('skip_trivial_constraints', ConfigValue(
+        default=False,
+        domain=bool,
+        description='Skip writing constraints whose body is constant'
+    ))
+    CONFIG.declare('file_determinism', ConfigValue(
+        default=FileDeterminism.ORDERED,
+        domain=InEnum(FileDeterminism),
+        description='How much effort to ensure file is deterministic',
+        doc="""
+        How much effort do we want to put into ensuring the
+        NL file is written deterministically for a Pyomo model:
+            NONE (0) : None
+            ORDERED (1): rely on underlying component ordering (default)
+            SORT_INDICES (2) : sort keys of indexed components
+            SORT_SYMBOLS (3) : sort keys AND sort names (over declaration order)
+        """
+    ))
+    CONFIG.declare('symbolic_solver_labels', ConfigValue(
+        default=False,
+        domain=bool,
+        description='Write the corresponding .row and .col files',
+    ))
+    CONFIG.declare('export_nonlinear_variables', ConfigValue(
```

Comment:
Should this be renamed? Maybe something like `export_unused_variables`?

Reply:
I would love to change this name.  It is here for backwards compatibility.

Reply:
Oh. I've never used that option. I didn't realize it existed. Another reason to add an alias feature to the config system...

Reply:
Agreed.  FWIW, I think this was added for PyNumero and ExternalGreyBoxModel: you have variables that only appear in external functions, so you need to "force" the NL writer to include variables that otherwise look fixed to the writer.